huggingface_mae.py

```diff
@@ -0,0 +1,339 @@
+        """Computes final loss and returns specific values of component losses for metric reporting."""
+        loss_dict = {}
+        target = self.input_norm(img)
+
+        reconstruction_ = unflatten_tokens(
+            reconstruction,
+            patch_size=self.patch_size,
+            num_modalities=self.in_chans if self.encoder.channel_agnostic else 1,
+        )
+
+        loss: torch.Tensor = self.loss(reconstruction_, target)
+
+        if isinstance(loss, list):
+            raise ValueError(
+                "Unmasked loss needs to return a tensor."
+                "You may need to set reduction to 'mean' or 'sum' in your loss function."
+            )
+
+        # Take the mean of the loss if it is not already reduced to make it a scalar
+        if loss.ndim > 1:
+            loss = loss.mean()
+        loss_dict[self.RECON_LOSS] = loss.item()
+
+        # compute fourier loss
+        if self.fourier_loss_weight > 0:
+            floss: torch.Tensor = self.fourier_loss(reconstruction_, target)
+            if floss.ndim > 1:
+                floss = floss.mean()
+
+            floss = floss * self.fourier_loss_weight
+            loss_dict[self.FOURIER_LOSS] = floss.item()
+
+        # here we use a mixing factor to keep the loss magnitude appropriate with fourier
+        if self.fourier_loss_weight > 0:
+            loss = loss + floss
+        return loss, loss_dict
+
+    def training_step(self, batch: TensorDict, batch_idx: int) -> TensorDict:
+        img = batch["pixels"]
+        latent, reconstruction, mask = self(img.clone())
+        # TODO: support configuration such that some losses are applied masked and some are not.
+        if self.apply_loss_unmasked:
+            full_loss, loss_dict = self.compute_unmasked_loss(
+                reconstruction, img.float()
+            )
+        else:
+            full_loss, loss_dict = self.compute_MAE_loss(
+                reconstruction, img.float(), mask
+            )
+        return {
+            "loss": full_loss,
+            **loss_dict,  # type: ignore[dict-item]
+        }
+
+    def validation_step(self, batch: TensorDict, batch_idx: int) -> TensorDict:
+        return self.training_step(batch, batch_idx)
+
+    def update_metrics(self, outputs: TensorDict, batch: TensorDict) -> None:
+        self.metrics["lr"].update(value=self.lr_scheduler.get_last_lr())
+        for key, value in outputs.items():
+            if key.endswith("loss"):
+                self.metrics[key].update(value)
+
+    def on_validation_batch_end(  # type: ignore[override]
+        self,
+        outputs: TensorDict,
+        batch: TensorDict,
+        batch_idx: int,
+        dataloader_idx: int = 0,
+    ) -> None:
+        super().on_validation_batch_end(outputs, batch, batch_idx, dataloader_idx)
+
+    def predict(self, imgs: torch.Tensor) -> torch.Tensor:
+        imgs = self.input_norm(imgs)
+        latent = self.encoder.forward(imgs)
+        return latent
```

Comment:
CC @okraus this is where we should explore other approaches to improve the Ph-Beta embeddings:
- using `self.encoder_decoder_proj(latent)`
- concatenating token embeddings for the different channels

It seems like doing encoder.forward here will call `self.vit_backbone.forward_head(x)` which I think does the `global_pool` step.

In any case, I would say we can try this:
```
imgs = self.input_norm(imgs)
tokens = self.encoder.vit_backbone.forward_features(imgs)

# options
if self.apply_decoder_projection_at_inference_time:  # can add as hparam to the model config
    tokens = self.encoder_decoder_proj(tokens)

if self.split_tokens_by_channel:  # can add as hparam to the model config
    tokens_per_channel = 256  # phenom-beta has 256 tokens per channel
    N, num_tokens, d = X.shape
    num_channels = num_tokens // tokens_per_channel  
    X_reshaped = X.view(N, num_channels, tokens_per_channel, d)
    pooled_segments = X_reshaped.mean(dim=2)  # Resulting shape: (N, num_channels, d)
    x = pooled_segments.view(N, num_channels * d).contiguous()
else:
    x = self.pool(tokens)
return x
```